File: Data Structures & Algorithms/daily-temperatures/submission-6.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int [] res = new int[n];
        Deque<Integer> st = new ArrayDeque<>();
        for(int i = 0; i < n;i++){
            int t = temperatures[i];
            while(!st.isEmpty() && t > temperatures[st.peek()]){
                int top = st.pop();
                res[top] = i - top;
            }
            st.push(i);
        }
        return res;
    }
}

File: Data Structures & Algorithms/asteroid-collision/submission-6.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for x in asteroids:
            if x > 0:
                stack.append(x)
                continue
            while stack and stack[-1] > 0:
                top = stack[-1]
                if top <= -x:
                    stack.pop()
                if top >= -x:
                    break
            else:
                stack.append(x)
        return stack